packages/affiliates/gas-rebate/VoterGasRebateV2.ts

```diff
@@ -0,0 +1,128 @@
+}
+
+export async function run(): Promise<void> {
+  console.log("Running UMA2.0 Gas rebate script! This script assumes you are running it for the previous month🍌.");
+
+  // Work out the range to run over. This should be over the totality of the previous month.
+  const currentDate = new Date();
+  const prevMonthStart = new Date(currentDate);
+  prevMonthStart.setMonth(prevMonthStart.getMonth() - 1);
+  prevMonthStart.setDate(1);
+  prevMonthStart.setHours(0, 0, 0, 0);
+
+  const prevMonthEnd = new Date(prevMonthStart.getFullYear(), prevMonthStart.getMonth() + 1, 0);
+  prevMonthEnd.setHours(23, 59, 59);
+
+  // Fetch associated block numbers for the start and end of the previous month.
+  const fromBlock = (await findBlockNumberAtTimestamp(getWeb3(), prevMonthStart.getTime() / 1000)).blockNumber;
+  const toBlock = (await findBlockNumberAtTimestamp(getWeb3(), prevMonthEnd.getTime() / 1000)).blockNumber;
+
+  console.log("Current time:", moment(currentDate).format());
+  console.log("Previous Month Start:", moment(prevMonthStart).format(), "& block", fromBlock);
+  console.log("Previous Month End:", moment(prevMonthEnd).format(), "& block", toBlock);
+
+  // Fetch all commit and reveal events.
+  const voting = await hre.ethers.getContractAt("VotingV2", await getAddress("VotingV2", 1));
+  const commitEvents = await voting.queryFilter(voting.filters.VoteCommitted(), fromBlock, toBlock);
+  const revealEvents = await voting.queryFilter(voting.filters.VoteRevealed(), fromBlock, toBlock);
+
+  // For each event find the associated transaction. We want to refund all transactions that were sent by voters.
+  const transactionsToRefund = await Promise.all(
+    [
+      commitEvents.map(async (commit: any) => voting.provider.getTransactionReceipt(commit.transactionHash)),
+      revealEvents.map(async (reveal: any) => voting.provider.getTransactionReceipt(reveal.transactionHash)),
+    ].flat()
+  );
+
+  console.log(
+    `In aggregate, refund ${commitEvents.length} commits and ${revealEvents.length} Reveals` +
+      ` for a total of ${transactionsToRefund.length} transactions`
+  );
+
+  // Find the associated UMA/ETH price for each transaction. We use the UniswapV3 pool to find the price at the block.
+  // Note that we search for 10000 blocks before to ensure that the range over which we have data spans the entire range.
+  const uniswapPool = await hre.ethers.getContractAt("UniswapV3", "0x157dfa656fdf0d18e1ba94075a53600d81cb3a97");
+  const uniswapPoolEvents = await uniswapPool.queryFilter(uniswapPool.filters.Swap(), fromBlock - 10000, toBlock);
+
+  const shareholderPayoutBN: { [key: string]: BigNumber } = {};
+  // Now, traverse all transactions and calculate the rebate for each.
+  for (const transaction of transactionsToRefund) {
+    // Eth used is the gas used * the gas price.
+    const ethUsed = transaction.gasUsed.mul(transaction.effectiveGasPrice);
```

Comment:
ye this is a great point and is something that I chose to ignore otherwise it would become really complex. one could use this to abuse the system...commit, reveal, or even call other external contracts all in the same function call to try and get free money from the system.

I think that's ok though. we will notice if there is an outlier and then we can update it. I did not want to over index on the complexity of this script.